src/main/java/solrocr/OcrHighlighter.java

```diff
@@ -394,92 +416,105 @@ public OcrHighlightResult[] highlightOcrFields(
+                  throw e;
+                } catch (IOException | RuntimeException e) {
+                  // This catch-all prevents OCR highlighting from failing the complete query,
+                  // instead users get an error message in their Solr log.
+                  if (contentFinal.getPointer() != null) {
+                    log.error(
+                        "Could not highlight OCR content for document {} at '{}'",
+                        docIdFinal,
+                        contentFinal.getPointer(),
+                        e);
+                  } else {
+                    log.error(
+                        "Could not highlight OCR for document {} with OCR markup '{}...'",
+                        docIdFinal,
+                        contentFinal.subSequence(0, 256),
+                        e);
+                  }
+                } finally {
+                  if (contentFinal instanceof AutoCloseable) {
+                    try {
+                      ((AutoCloseable) contentFinal).close();
+                    } catch (Exception e) {
+                      log.warn(
+                          "Encountered error while closing content iterator for {}: {}",
+                          contentFinal.getPointer(),
+                          e.getMessage());
+                    }
+                  }
+                }
+              };
+          if (hlThreadPool.getActiveCount() == hlThreadPool.getMaximumPoolSize()) {
+            // Run the task synchronously on the current thread if the pool is fully busy
+            try {
+              hlFn.run();
+            } catch (ExitingIterCharSeq.ExitingIterCharSeqException
+                | ExitableDirectoryReader.ExitingReaderException e) {
+              if (respHeader.get(PARTIAL_OCR_HIGHLIGHTS) == null) {
+                respHeader.put(PARTIAL_OCR_HIGHLIGHTS, Boolean.TRUE);
+                log.warn("OCR Highlighting timed out while handling " + content.getPointer(), e);
               }
+              resultByDocIn[docInIndex] = null;
             }
+          } else {
+            // Otherwise speed up highlighting by parallelizing the work as much as possible
+            hlFuts.add(hlThreadPool.submit(hlFn));
           }
-          snippetCountsByField[fieldIdx][docInIndex] = fieldHighlighter.getNumMatches(docId);
         }
       }
       batchDocIdx += fieldValsByDoc.size();
     }
     assert docIdIter.docID() == DocIdSetIterator.NO_MORE_DOCS
         || docIdIter.nextDoc() == DocIdSetIterator.NO_MORE_DOCS;
 
+    if (!hlFuts.isEmpty()) {
+      boolean partialOcrHighlights = false;
+      for (Future<?> fut : hlFuts) {
+        try {
+          fut.get();
+        } catch (ExecutionException e) {
+          if (e.getCause() instanceof ExitingIterCharSeq.ExitingIterCharSeqException
+              || e.getCause() instanceof ExitableDirectoryReader.ExitingReaderException) {
+            partialOcrHighlights = true;
```

Comment:
Similar to the comment above: Should we cancel the other Futures after a timeout has occurred?
When sequential highlighting fails, we explicitly set `resultByDocIn[docInIndex] = null;`. Do we have to do this in the parallel case as well?

Reply:
> Should we cancel the other Futures after a timeout has occurred?

In this case, no. We're not iterating the futures in their order of completion, i.e. just because one task encountered a timeout, doesn't mean all subsequent tasks in the list have failed as well. If the timed out task took a long time, there might be a lot of completed tasks still left, that's why we check those as well and use the flag to keep track of timed out tasks.

> When sequential highlighting fails, we explicitly set resultByDocIn[docInIndex] = null;. Do we have to do this in the parallel case as well?

Yes, I noticed this as well, I fixed it to handle this in the catch-block inside of the thread already.

Reply:
I was just wondering: We wouldn't want to discard the Futures that already completed successfully, but should we cancel those Futures that are still running or pending if one Future has timed out?
Wouldn't the timeout build up otherwise due to the time the tasks spend waiting in the queue?

Reply:
Ah, I think I get it now: `ExitingIterCharSeq content` is shared by all threads, so the timeout is tracked per field content.